# Project 10 -- SEYI OGUNMODEDE

## Question 1

In [1]:
options(jupyter.rich_display = F)

In [2]:
options(repr.matrix.max.cols=50, repr.matrix.max.rows=200)

In [3]:
library(data.table)

In [4]:
users<- fread("/anvil/projects/tdm/data/okcupid/filtered/users.csv")

# fread, though will determine automatically what the dome meter are in the file

In [5]:
questions <- fread("/anvil/projects/tdm/data/okcupid/filtered/questions.csv")
# fread though will determine automatically what the dome meter are in the file

In [6]:
dim (users)

[1] 68371  2284

In [7]:
dim(questions)

[1] 2281   10

In [8]:
head(users)

  q41                  q42 q43 q45 q46         q48   q49      q50 q55 q56 q57
1 NA                   NA  NA  NA  NA          NA    Carefree NA  No  No  NA 
2 NA                   NA  NA  NA  NA          NA    NA       NA  NA  NA  NA 
3 Extremely important  NA  NA  NA  Good        NA    Carefree NA  No  NA  NA 
4 NA                   NA  NA  NA  NA          NA    NA       NA  NA  NA  NA 
5 Not at all important NA  NA  NA  Interesting Weird Intense  NA  NA  NA  NA 
6 Not at all important NA  NA  NA  Good        Weird NA       NA  NA  No  No 
  q60          q61 q63 q65 q67 q68 q69 q70 q71 q73      q74     q76 q77      
1 NA           NA  NA  NA  NA  NA  NA  No  No  NA       NA      NA  Sometimes
2 NA           NA  NA  NA  NA  NA  NA  NA  NA  NA       NA      NA  NA       
3 Warm-hearted NA  NA  NA  NA  NA  NA  No  No  Big city Clothes NA  Rarely   
4 NA           NA  NA  NA  NA  NA  NA  NA  NA  NA       NA      NA  Sometimes
5 NA           NA  NA  NA  NA  NA  NA  No  NA  NA       NA      

In [9]:
head(questions)

  X  
1 q41
2 q42
3 q43
4 q44
5 q45
6 q46
  text                                                                                               
1 How important is religion/God in your life?                                                        
2 Is your duty to religion/God the most important thing in your life?                                
3 Do you believe in some kind of a spiritual, cosmic force outside the realm of most modern religion?
4 Some religions are more correct than others.                                                       
5 Do you believe in ghosts?                                                                          
6 Would you prefer good things happened, or interesting things?                                      
  option_1            option_2           option_3          
1 Extremely important Somewhat important Not very important
2 Yes                 No                                   
3 Yes                 No                                   
4 Tr

In [10]:
head(users$gender_orientation)

[1] "Hetero_female"   "Hetero_male"     "Hetero_female"   "Hetero_female"  
[5] "Bisexual_female" "Hetero_male"    

In [11]:
tail(users$gender_orientation)

[1] "Hetero_male"   "Hetero_male"   "Hetero_female" "Hetero_male"  
[5] "Hetero_female" "Hetero_male"  

In [12]:
head(users$race)

[1] "White" NA      NA      "White" NA      "White"

In [13]:
tail(users$race)

[1] "White" "White" NA      NA      "White" "Other"

In [14]:
head(questions$Type)

[1] "O" "O" "O" "O" "O" "N"

In [15]:
tail(questions$Type)

[1] "" "" "" "" "" ""

In [16]:
head(questions$text)

[1] "How important is religion/God in your life?"                                                        
[2] "Is your duty to religion/God the most important thing in your life?"                                
[3] "Do you believe in some kind of a spiritual, cosmic force outside the realm of most modern religion?"
[4] "Some religions are more correct than others."                                                       
[5] "Do you believe in ghosts?"                                                                          
[6] "Would you prefer good things happened, or interesting things?"                                      

In [17]:
tail(questions$text)

[1] "Type of match"     "Max age of match"  "Looking for match"
[4] "Location of match" "Min age of match"  "Status of match"  

We want to go ahead and load the datasets into data.frames named users and questions. 

Take a look at both data.frames and identify what is a part of each of them. 
#user data frame has 68371 rows  2284 columns
#questions data frame has 2281 rows 10 columns

What information is in each datatset, and how they are related?

For the the user, it is a data set about human being in respect to their gender and what they are interested in.
for the question, it is a data set that has series of questions related to the user, in respect to texs, type, options, keywords as it applies.


## Question 2

In [18]:
?grep
# grep helps to look for pattern inside text
# to see help function using grep
# to see other function the help using grep (in different ways to use grep)

grep                   package:base                    R Documentation

_P_a_t_t_e_r_n _M_a_t_c_h_i_n_g _a_n_d _R_e_p_l_a_c_e_m_e_n_t

_D_e_s_c_r_i_p_t_i_o_n:

     'grep', 'grepl', 'regexpr', 'gregexpr', 'regexec' and 'gregexec'
     search for matches to argument 'pattern' within each element of a
     character vector: they differ in the format of and amount of
     detail in the results.

     'sub' and 'gsub' perform replacement of the first and all matches
     respectively.

_U_s_a_g_e:

     grep(pattern, x, ignore.case = FALSE, perl = FALSE, value = FALSE,
          fixed = FALSE, useBytes = FALSE, invert = FALSE)
     
     grepl(pattern, x, ignore.case = FALSE, perl = FALSE,
           fixed = FALSE, useBytes = FALSE)
     
     sub(pattern, replacement, x, ignore.case = FALSE, perl = FALSE,
         fixed = FALSE, useBytes = FALSE)
     
     gsub(pattern, replacement, x, ignore.case = FALSE, perl = FALSE,
          fixed = FALSE

In [19]:
dim(questions)
# it has 10 columns

[1] 2281   10

In [20]:
head(questions)
# to see the word gooogle, it may be possible with text

  X  
1 q41
2 q42
3 q43
4 q44
5 q45
6 q46
  text                                                                                               
1 How important is religion/God in your life?                                                        
2 Is your duty to religion/God the most important thing in your life?                                
3 Do you believe in some kind of a spiritual, cosmic force outside the realm of most modern religion?
4 Some religions are more correct than others.                                                       
5 Do you believe in ghosts?                                                                          
6 Would you prefer good things happened, or interesting things?                                      
  option_1            option_2           option_3          
1 Extremely important Somewhat important Not very important
2 Yes                 No                                   
3 Yes                 No                                   
4 Tr

In [21]:
grep("Google",questions$text)
# it does find it one time on row 2172

[1] 2172

In [22]:
questions[2172, ]
# to detect what is in the column

  X       text                                       option_1                
1 q170849 Do you Google someone before a first date? Yes. Knowledge is power!
  option_2                   option_3 option_4 N     Type Order
1 No. Why spoil the mystery?                   39621 O         
  Keywords               
1 descriptive; technology

In [23]:
questions$text[grepl("Google",questions$text)]
# another way is using logical grep (grepl)
# a neat thing about grepl is use the location at the begining as a position within the vector of text within the text column

[1] "Do you Google someone before a first date?"

What do you notice if you just use the function grep() and create a new variable google and then print that variable?
 it helps to locate the number of occurence and the row it appears.

Now that you know the row number, how can you take a look at the information there?

  X       text                                       option_1                
1 q170849 Do you Google someone before a first date? Yes. Knowledge is power!
  option_2                   option_3 option_4 N     Type Order
1 No. Why spoil the mystery?                   39621 O         
  Keywords               
1 descriptive; technology

using grepl

"Do you Google someone before a first date?"


## Question 3

In [24]:
dim(users)
# 68371rows  2284columns
# roughly one colmn per question

[1] 68371  2284

In [25]:
head(users)
# showing one column per question

  q41                  q42 q43 q45 q46         q48   q49      q50 q55 q56 q57
1 NA                   NA  NA  NA  NA          NA    Carefree NA  No  No  NA 
2 NA                   NA  NA  NA  NA          NA    NA       NA  NA  NA  NA 
3 Extremely important  NA  NA  NA  Good        NA    Carefree NA  No  NA  NA 
4 NA                   NA  NA  NA  NA          NA    NA       NA  NA  NA  NA 
5 Not at all important NA  NA  NA  Interesting Weird Intense  NA  NA  NA  NA 
6 Not at all important NA  NA  NA  Good        Weird NA       NA  NA  No  No 
  q60          q61 q63 q65 q67 q68 q69 q70 q71 q73      q74     q76 q77      
1 NA           NA  NA  NA  NA  NA  NA  No  No  NA       NA      NA  Sometimes
2 NA           NA  NA  NA  NA  NA  NA  NA  NA  NA       NA      NA  NA       
3 Warm-hearted NA  NA  NA  NA  NA  NA  No  No  Big city Clothes NA  Rarely   
4 NA           NA  NA  NA  NA  NA  NA  NA  NA  NA       NA      NA  Sometimes
5 NA           NA  NA  NA  NA  NA  NA  No  NA  NA       NA      

In [26]:
which(names(users) == "q170849")
# to look for which one is TRUE
#out of the users  data frame, colmn 1193 is TRUE

[1] 1193

In [27]:
users[ ,1193]
# to detect what is in the row

      q170849                   
1     No. Why spoil the mystery?
2     NA                        
3     No. Why spoil the mystery?
4     NA                        
5     NA                        
6     Yes. Knowledge is power!  
7     Yes. Knowledge is power!  
8     NA                        
9     NA                        
10    NA                        
11    No. Why spoil the mystery?
12    NA                        
13    Yes. Knowledge is power!  
14    NA                        
15    No. Why spoil the mystery?
16    NA                        
17    NA                        
18    NA                        
19    No. Why spoil the mystery?
20    NA                        
21    NA                        
22    No. Why spoil the mystery?
23    Yes. Knowledge is power!  
24    Yes. Knowledge is power!  
25    NA                        
26    No. Why spoil the mystery?
27    NA                        
28    NA                        
29    NA                        
30    No. 

In [28]:
table(users[[1193]], useNA = "always")
# another way to look at it
# because the outcome contain NA's, 
# then use table and use NA to know how many NA are there


No. Why spoil the mystery?   Yes. Knowledge is power! 
                     25804                      13817 
                      <NA> 
                     28750 

In [29]:
prop.table(table(users[[1193]], useNA = "always"))
# this shows the proportion


No. Why spoil the mystery?   Yes. Knowledge is power! 
                 0.3774115                  0.2020886 
                      <NA> 
                 0.4204999 

In [30]:
(prop.table(table(users[[1193]], useNA = "always")))*100
# to know the actual percentage by multiplying by 100


No. Why spoil the mystery?   Yes. Knowledge is power! 
                  37.74115                   20.20886 
                      <NA> 
                  42.04999 

In [31]:
users[2172 ,1193]
# for verification

  q170849                   
1 No. Why spoil the mystery?

Using the row from our previous question, which variable does this correspond with in the data.frame users?
 colmn 1193
Row 2172 in questions corresponds to column named q170849 in users

Knowing that the two possible answers are "No. Why spoil the mystery?" and "Yes, Knowledge is power!" What percentage of users do NOT google someone before the first date? 37.74%

## Question 4

In [ ]:
tapply(users[[1193]], gender, table)
# we want to break things up using the table 
# we are yet to look at the gender
# colomn to consider that is for gender is (gender 2 in column 2282)

In [33]:
head(users[[2282]])
# these are the genders there.
# the column named gender2 is 2282


[1] "Woman" "Man"   "Woman" "Woman" "Woman" "Man"  

In [34]:
tapply(users[[1193]], users[[2282]], table)
# we are going to take the data from how people responded to the Google questions
# break it up according to the gender
# within all the males responses, we take the data and make a table
# within all the females responses, we take the data and make a table


$Man

No. Why spoil the mystery?   Yes. Knowledge is power! 
                     17294                       7549 

$Woman

No. Why spoil the mystery?   Yes. Knowledge is power! 
                      7987                       5660 


In [35]:
tapply(users[[1193]], users[[2282]], table,useNA = "always")
# put useNA = "always" to know the number of people that did not respond in each gender

$Man

No. Why spoil the mystery?   Yes. Knowledge is power! 
                     17294                       7549 
                      <NA> 
                     15372 

$Woman

No. Why spoil the mystery?   Yes. Knowledge is power! 
                      7987                       5660 
                      <NA> 
                     12305 


In [36]:
tapply(users[[1193]], users[[2282]], function(x) {prop.table(table(x,useNA = "always"))})
# another way to calculate the percentage using prop.table

$Man
x
No. Why spoil the mystery?   Yes. Knowledge is power! 
                 0.4300385                  0.1877160 
                      <NA> 
                 0.3822454 

$Woman
x
No. Why spoil the mystery?   Yes. Knowledge is power! 
                 0.3077605                  0.2180949 
                      <NA> 
                 0.4741446 


In [37]:
tapply(users[[1193]], users[[2282]], function(x) {(prop.table(table(x,useNA = "always")))*100})

$Man
x
No. Why spoil the mystery?   Yes. Knowledge is power! 
                  43.00385                   18.77160 
                      <NA> 
                  38.22454 

$Woman
x
No. Why spoil the mystery?   Yes. Knowledge is power! 
                  30.77605                   21.80949 
                      <NA> 
                  47.41446 


In [38]:
sapply(tapply(users[[1193]], users[[2282]], table,useNA = "always"),prop.table)
# this apply prop.table to each
# this take data from men and make proportion to women

                           Man       Woman    
No. Why spoil the mystery? 0.4300385 0.3077605
Yes. Knowledge is power!   0.1877160 0.2180949
<NA>                       0.3822454 0.4741446

In [ ]:
# sapply works like this:    sapply(mydata, myfunction)
# sapply takes some data and a function that you want to run on each piece of the data

Using the ability to create a function AND tapply find the percentages of Female vs Male (Man vs Woman, as categorized in the users data.frame) who DO google someone before their date.

## Question 5

In [39]:
head(users$gender_orientation)

[1] "Hetero_female"   "Hetero_male"     "Hetero_female"   "Hetero_female"  
[5] "Bisexual_female" "Hetero_male"    

In [40]:
head(users$gender_orientation, n=20)

 [1] "Hetero_female"   "Hetero_male"     "Hetero_female"   "Hetero_female"  
 [5] "Bisexual_female" "Hetero_male"     NA                "Hetero_male"    
 [9] "Hetero_female"   "Hetero_male"     "Hetero_female"   "Hetero_male"    
[13] "Bisexual_female" "Hetero_female"   "Hetero_male"     "Bisexual_female"
[17] NA                "Hetero_female"   "Hetero_male"     "Hetero_female"  

In [41]:
gsub("_", "", head(users$gender_orientation, n=20))
# take the underscore out and everything after underscore
# to do that, i do a global substitution of the underscore and put in nothing in its place
# it will just take out the underscore itself
# but it wont get things after the underscore
# lots of way to solve the problems

 [1] "Heterofemale"   "Heteromale"     "Heterofemale"   "Heterofemale"  
 [5] "Bisexualfemale" "Heteromale"     NA               "Heteromale"    
 [9] "Heterofemale"   "Heteromale"     "Heterofemale"   "Heteromale"    
[13] "Bisexualfemale" "Heterofemale"   "Heteromale"     "Bisexualfemale"
[17] NA               "Heterofemale"   "Heteromale"     "Heterofemale"  

In [42]:
gsub("_[A-Za-z]*", "", head(users$gender_orientation, n=20))
# i will take any letter that comes after that
# [A-Za-z] means take any letter capitals or smallletters that comes after underscore
# * means  take as many of those you find (by putting *)
#  "" means substituting them out to make them into nothing

 [1] "Hetero"   "Hetero"   "Hetero"   "Hetero"   "Bisexual" "Hetero"  
 [7] NA         "Hetero"   "Hetero"   "Hetero"   "Hetero"   "Hetero"  
[13] "Bisexual" "Hetero"   "Hetero"   "Bisexual" NA         "Hetero"  
[19] "Hetero"   "Hetero"  

In [43]:
gsub("_[A-Za-z]*", "burger", head(users$gender_orientation, n=20))
# the thing you put in here ("") is what get sustituted
# if you put burger in it (e.g "burger")

 [1] "Heteroburger"   "Heteroburger"   "Heteroburger"   "Heteroburger"  
 [5] "Bisexualburger" "Heteroburger"   NA               "Heteroburger"  
 [9] "Heteroburger"   "Heteroburger"   "Heteroburger"   "Heteroburger"  
[13] "Bisexualburger" "Heteroburger"   "Heteroburger"   "Bisexualburger"
[17] NA               "Heteroburger"   "Heteroburger"   "Heteroburger"  

In [44]:
gsub("_[A-Za-z]*", "", head(users$gender_orientation, n=20))
# we want to remove everything after the underscore

 [1] "Hetero"   "Hetero"   "Hetero"   "Hetero"   "Bisexual" "Hetero"  
 [7] NA         "Hetero"   "Hetero"   "Hetero"   "Hetero"   "Hetero"  
[13] "Bisexual" "Hetero"   "Hetero"   "Bisexual" NA         "Hetero"  
[19] "Hetero"   "Hetero"  

In [45]:
table(gsub("_[A-Za-z]*", "", head(users$gender_orientation, n=20)))
# then make table of the result


Bisexual   Hetero 
       3       15 

In [46]:
table(gsub("_[A-Za-z]*", "", head(users$gender_orientation, n=100)))
# table for the first 100 entries


Bisexual      Gay   Hetero 
      12        4       78 

In [47]:
table(gsub("_[A-Za-z]*", "", users$gender_orientation), useNA="always")
# since i have gotten for first 100 entries
# for all together i will take the head off, and the range(n=100)
# lastly you could check, is there any NA's in there (useNA="always")


Bisexual      Gay   Hetero     <NA> 
    4743     3392    57747     2489 

Using the ability to create a function AND use sapply() write a function that takes the string and removes everything after/including the _ from the gender_orientation column in the users data.frame.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.